jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/DefaultCredentialRetrievers.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2018 Google LLC. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.plugins.common;
+
+import com.google.cloud.tools.jib.configuration.credentials.Credential;
+import com.google.cloud.tools.jib.configuration.credentials.CredentialRetriever;
+import com.google.cloud.tools.jib.frontend.CredentialRetrieverFactory;
+import com.google.cloud.tools.jib.registry.credentials.DockerCredentialHelperFactory;
+import java.util.ArrayList;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * Generates a list of default {@link CredentialRetriever}s.
+ *
+ * <p>The retrievers are, in order of first-checked to last-checked:
+ *
+ * <ol>
+ *   <li>{@link CredentialRetrieverFactory#dockerCredentialHelper} for a known credential helper, if
+ *       set
+ *   <li>{@link CredentialRetrieverFactory#known} for known credential, if set
+ *   <li>{@link CredentialRetrieverFactory#inferCredentialHelper}
+ *   <li>{@link CredentialRetrieverFactory#dockerConfig}
+ * </ol>
+ */
+public class DefaultCredentialRetrievers {
+
+  /**
+   * Creates a new {@link DefaultCredentialRetrievers} with a given {@link
+   * CredentialRetrieverFactory}.
+   *
+   * @param credentialRetrieverFactory the {@link CredentialRetrieverFactory} to generate the {@link
+   *     CredentialRetriever}s
+   * @return a new {@link DefaultCredentialRetrievers}
+   */
+  public static DefaultCredentialRetrievers init(
```

Comment:
Why not just use a public constructor?

Reply:
I'm intending this class to be potentially used by users of Jib library (this may be moved into `jib-core`). For all public APIs, I'm trying to keep it so that there are no constructors exposed (ie. static factories over constructors).